File: pkg/fenster/src/fenster_mac.m
#include "fenster.h"
#import <Cocoa/Cocoa.h>

static const uint8_t FENSTER_KEYCODES[128] = {65,83,68,70,72,71,90,88,67,86,0,66,81,87,69,82,89,84,49,50,51,52,54,53,61,57,55,45,56,48,93,79,85,91,73,80,10,76,74,39,75,59,92,44,47,78,77,46,9,32,96,8,0,27,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,26,2,3,127,0,5,0,4,0,20,19,18,17,0};

@interface FensterDelegate : NSObject <NSWindowDelegate>
- (void)windowDidResize:(NSNotification *)notification;
@end

@interface FensterView : NSView
@end

@implementation FensterDelegate

- (BOOL)windowShouldClose:(NSWindow *)sender {
    [NSApp terminate:nil];
    return YES;
}

- (void)windowDidResize:(NSNotification *)notification {
    NSWindow *window = notification.object;
    FensterView *view = (FensterView *)window.contentView;
    fenster *f = (__bridge fenster *)objc_getAssociatedObject(view, @selector(fenster));

    NSRect frame = [view frame];
    f->width = frame.size.width;
    f->height = frame.size.height;
    f->buf = f->realloc(
        f->user_ctx, f->buf, f->width * f->height * sizeof(uint32_t)); 

    [view setNeedsDisplay:YES];
}

@end

@implementation FensterView

- (void)drawRect:(NSRect)dirtyRect {
    fenster* f = (__bridge fenster*)objc_getAssociatedObject(self, @selector(fenster));
    CGContextRef context = [[NSGraphicsContext currentContext] CGContext];
    CGColorSpaceRef space = CGColorSpaceCreateDeviceRGB();
    CGDataProviderRef provider = CGDataProviderCreateWithData(NULL, f->buf, f->width * f->height * sizeof(uint32_t), NULL);
    CGImageRef img = CGImageCreate(f->width, f->height, 8, 32, f->width * 4, space,
                                   kCGImageAlphaNoneSkipFirst | kCGBitmapByteOrder32Little,
                                   provider, NULL, false, kCGRenderingIntentDefault);
    CGColorSpaceRelease(space);
    CGDataProviderRelease(provider);
    CGContextDrawImage(context, CGRectMake(0, 0, f->width, f->height), img);
    CGImageRelease(img);
}

@end

int fenster_open(fenster* f) {
  f->buf = f->realloc(
      f->user_ctx, f->buf, f->width * f->height * sizeof(uint32_t)); 
  [NSApplication sharedApplication];
  [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
  
  f->platform.wnd = [[NSWindow alloc]
    initWithContentRect:NSMakeRect(0, 0, f->width, f->height)
    styleMask:
      NSWindowStyleMaskTitled |
      NSWindowStyleMaskClosable |
      NSWindowStyleMaskResizable |
      NSWindowStyleMaskMiniaturizable
    backing:NSBackingStoreBuffered
    defer:NO];

  FensterDelegate* delegate = [[FensterDelegate alloc] init];
  [f->platform.wnd setDelegate:delegate];
  
  FensterView* view = [[FensterView alloc] init];
  [f->platform.wnd setContentView:view];
  objc_setAssociatedObject(
      view,
      @selector(fenster),
      (__bridge id)f, OBJC_ASSOCIATION_ASSIGN);
  
  [f->platform.wnd setTitle:[NSString stringWithUTF8String:f->title]];
  [f->platform.wnd makeKeyAndOrderFront:nil];
  [f->platform.wnd center];
  [NSApp activateIgnoringOtherApps:YES];
  
  return 0;
}

void fenster_close(fenster* f) {
  [f->platform.wnd close];
  f->buf = f->realloc(f->user_ctx, f->buf, 0);
}

void fenster_paint(fenster* f) {
  [[f->platform.wnd contentView] setNeedsDisplay:YES];
}

int fenster_loop(fenster* f) {
  NSEvent* event = [NSApp
    nextEventMatchingMask:NSEventMaskAny
    untilDate:nil
    inMode:NSDefaultRunLoopMode
    dequeue:YES];
  if (!event) return 0;

  switch ([event type]) {
      case NSEventTypeLeftMouseDown:
          f->mouse = FENSTER_LMOUSE_DOWN;
          break;
      case NSEventTypeLeftMouseUp:
          f->mouse = FENSTER_LMOUSE_UP;
          break;
      case NSEventTypeRightMouseDown:
          f->mouse = FENSTER_RMOUSE_DOWN;
          break;
      case NSEventTypeRightMouseUp:
          f->mouse = FENSTER_RMOUSE_UP;
          break;
      case NSEventTypeMouseMoved:
      case NSEventTypeLeftMouseDragged: {
          NSPoint point = [event locationInWindow];
          f->x = (int)point.x;
          f->y = (int)(f->height - point.y);
          return 0;
      }
      case NSEventTypeKeyDown:
      case NSEventTypeKeyUp: {
          NSUInteger keyCode = [event keyCode];
          f->keys[keyCode < 127 ? FENSTER_KEYCODES[keyCode] : 0] =
              [event type] == NSEventTypeKeyDown ? 1 : -1;
          NSUInteger modifiers = [event modifierFlags] >> 17;
          f->mod = (modifiers & 0xc)
            | ((modifiers & 1) << 1)
            | ((modifiers >> 1) & 1);
          return 0;
      }
  }
  
  [NSApp sendEvent:event];
  return 0;
}

File: pkg/fenster/src/fenster_mac.c
#ifdef __APPLE__

#include "fenster.h"

#define msg(r, o, s) ((r(*)(id, SEL))objc_msgSend)(o, sel_getUid(s))
#define msg1(r, o, s, A, a)                                                    \
  ((r(*)(id, SEL, A))objc_msgSend)(o, sel_getUid(s), a)
#define msg2(r, o, s, A, a, B, b)                                              \
  ((r(*)(id, SEL, A, B))objc_msgSend)(o, sel_getUid(s), a, b)
#define msg3(r, o, s, A, a, B, b, C, c)                                        \
  ((r(*)(id, SEL, A, B, C))objc_msgSend)(o, sel_getUid(s), a, b, c)
#define msg4(r, o, s, A, a, B, b, C, c, D, d)                                  \
  ((r(*)(id, SEL, A, B, C, D))objc_msgSend)(o, sel_getUid(s), a, b, c, d)

#define cls(x) ((id)objc_getClass(x))

extern id const NSDefaultRunLoopMode;
extern id const NSApp;

static void fenster_draw_rect(id v, SEL s, CGRect r) {
  (void)r, (void)s;
  fenster *f = (fenster *)objc_getAssociatedObject(v, "fenster");
  CGContextRef context =
      msg(CGContextRef, msg(id, cls("NSGraphicsContext"), "currentContext"),
          "graphicsPort");
  CGColorSpaceRef space = CGColorSpaceCreateDeviceRGB();
  CGDataProviderRef provider = CGDataProviderCreateWithData(
      NULL, f->buf, f->width * f->height * 4, NULL);
  CGImageRef img =
      CGImageCreate(f->width, f->height, 8, 32, f->width * 4, space,
                    kCGImageAlphaNoneSkipFirst | kCGBitmapByteOrder32Little,
                    provider, NULL, false, kCGRenderingIntentDefault);
  CGColorSpaceRelease(space);
  CGDataProviderRelease(provider);
  CGContextDrawImage(context, CGRectMake(0, 0, f->width, f->height), img);
  CGImageRelease(img);
}

static BOOL fenster_should_close(id v, SEL s, id w) {
  (void)v, (void)s, (void)w;
  msg1(void, NSApp, "terminate:", id, NSApp);
  return YES;
}

int fenster_open(fenster *f) {
  msg(id, cls("NSApplication"), "sharedApplication");
  msg1(void, NSApp, "setActivationPolicy:", NSInteger, 0);
  f->platform.wnd = msg4(id, msg(id, cls("NSWindow"), "alloc"),
                "initWithContentRect:styleMask:backing:defer:", CGRect,
                CGRectMake(0, 0, f->width, f->height), NSUInteger, 3,
                NSUInteger, 2, BOOL, NO);
  Class windelegate =
      objc_allocateClassPair((Class)cls("NSObject"), "FensterDelegate", 0);
  class_addMethod(windelegate, sel_getUid("windowShouldClose:"),
                  (IMP)fenster_should_close, "c@:@");
  objc_registerClassPair(windelegate);
  msg1(void, f->platform.wnd, "setDelegate:", id,
       msg(id, msg(id, (id)windelegate, "alloc"), "init"));
  Class c = objc_allocateClassPair((Class)cls("NSView"), "FensterView", 0);
  class_addMethod(c, sel_getUid("drawRect:"), (IMP)fenster_draw_rect, "i@:@@");
  objc_registerClassPair(c);

  id v = msg(id, msg(id, (id)c, "alloc"), "init");
  msg1(void, f->platform.wnd, "setContentView:", id, v);
  objc_setAssociatedObject(v, "fenster", (id)f, OBJC_ASSOCIATION_ASSIGN);

  id title = msg1(id, cls("NSString"), "stringWithUTF8String:", const char *,
                  f->title);
  msg1(void, f->platform.wnd, "setTitle:", id, title);
  msg1(void, f->platform.wnd, "makeKeyAndOrderFront:", id, nil);
  msg(void, f->platform.wnd, "center");
  msg1(void, NSApp, "activateIgnoringOtherApps:", BOOL, YES);
  return 0;
}

void fenster_close(fenster *f) {
  msg(void, f->platform.wnd, "close");
}

// clang-format off
static const uint8_t FENSTER_KEYCODES[128] = {65,83,68,70,72,71,90,88,67,86,0,66,81,87,69,82,89,84,49,50,51,52,54,53,61,57,55,45,56,48,93,79,85,91,73,80,10,76,74,39,75,59,92,44,47,78,77,46,9,32,96,8,0,27,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,26,2,3,127,0,5,0,4,0,20,19,18,17,0};
// clang-format on

void fenster_paint(fenster* f) {
  msg1(void, msg(id, f->platform.wnd, "contentView"), "setNeedsDisplay:", BOOL, YES);
}

int fenster_loop(fenster *f) {
  id ev = msg4(id, NSApp,
               "nextEventMatchingMask:untilDate:inMode:dequeue:", NSUInteger,
               NSUIntegerMax, id, NULL, id, NSDefaultRunLoopMode, BOOL, YES);
  if (!ev)
    return 0;
  NSUInteger evtype = msg(NSUInteger, ev, "type");
  switch (evtype) {
  case 1: /* NSEventTypeMouseDown */
    f->mouse |= 1;
    break;
  case 2: /* NSEventTypeMouseUp*/
    f->mouse &= ~1;
    break;
  case 5:
  case 6: { /* NSEventTypeMouseMoved */
    CGPoint xy = msg(CGPoint, ev, "locationInWindow");
    f->x = (int)xy.x;
    f->y = (int)(f->height - xy.y);
    return 0;
  }
  case 10: /*NSEventTypeKeyDown*/
  case 11: /*NSEventTypeKeyUp:*/ {
    NSUInteger k = msg(NSUInteger, ev, "keyCode");
    f->keys[k < 127 ? FENSTER_KEYCODES[k] : 0] = evtype == 10;
    NSUInteger mod = msg(NSUInteger, ev, "modifierFlags") >> 17;
    f->mod = (mod & 0xc) | ((mod & 1) << 1) | ((mod >> 1) & 1);
    return 0;
  }
  }
  msg1(void, NSApp, "sendEvent:", id, ev);
  return 0;
}

#endif
